### ISSUE TITLE
(Auto)Build container next to sources so miners can deposit to them

### ISSUE THREAD

**Issue by cyberblast** [ScreepsOCS/screeps.behaviour-action-pattern#331]
_From @Dasku on November 8, 2016 21:40_

Build automation

_Copied from original issue: cyberblast/screeps.ocs.internal#33_

**Comment by ESchuderer** [comment]
Quick and dirty test, works quite good.

**ToDo:**
- Use a regular iterating for loop instead of "for ... of"
- Is it saving CPU to also check for construction sites?
- clean it up

```javascript
if(!memory.hasOwnProperty('container')) {
            let sources;
            memory.container = {};
            
            if (Memory.rooms[key] && Memory.rooms[key]) {
                sources = Memory.rooms[key].sources
            } else if(room) {
                sources = room.sources;
            }
            
            if(sources) {
                let object;
                let containers;
                let x;
                let y;
                
                for(let source of sources) {
                    object =  Game.getObjectById(source);
                    
                    if(object) {
                        x = object.pos.x;
                        y = object.pos.y;
                        
                        containers = room.lookForAtArea(LOOK_TERRAIN, y-1, x-1, y+1, x+1, true).filter(function(element) {
                            return element.terrain == "swamp" || element.terrain == "plain";
                        });
                        
                        memory.container[source] = {
                            x: containers[0].x,
                            y: containers[0].y
                        };
                    }
                }
            }
        } else if(room) {
            let exists;
            console.log(room);
            for(let container in memory.container) {
                console.log(container);
                exists = room.lookForAt(LOOK_STRUCTURES, memory.container[container].x, memory.container[container].y).filter(function(element) {
                    return element.structureType == STRUCTURE_CONTAINER;
                });
                
                if(exists.length < 1) {
                    room.createConstructionSite(memory.container[container].x, memory.container[container].y, STRUCTURE_CONTAINER);
                }
            }
        }
```

**Comment by chobobobo** [comment]
@ESchuderer I think a quick check for an existing container/container construction site adjacent to the source would be good too before we choose a new location in case the user has already placed one.  Also, I think mod.strategies.container would be a good place to put this code rather than in mod.memory (we want memory access to be quick).

**Comment by karlthepagan** [comment]
My only advice is: avoid putting code which will never be overridden by a different creep type in strategies.

I also think iterating on 2-6 room tiles looking for structures & construction sites is going to be more reliable (and posssibly more efficient at high scales).

**Comment by chobobobo** [comment]
This is implemented in #714
